File: sdk/examples/battleship/sawtooth_battleship/processor/handler.py
# Copyright 2016 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

import hashlib

from sawtooth_sdk.processor.state import StateEntry
from sawtooth_sdk.processor.exceptions import InvalidTransaction
from sawtooth_sdk.processor.exceptions import InternalError
from sawtooth_sdk.protobuf.transaction_pb2 import TransactionHeader

LOGGER = logging.getLogger(__name__)


class BattleshipHandler(object):
    def __init__(self, namespace_prefix):
        self._namespace_prefix = namespace_prefix

    @property
    def family_name(self):
        return 'battleship'

    @property
    def family_versions(self):
        return ['1.0']

    @property
    def encodings(self):
        return ['csv-utf8']

    @property
    def namespaces(self):
        return [self._namespace_prefix]


    def apply(self, transaction, state_store):

        # 1. Deserialize the transaction and verify it is valid
        header = TransactionHeader()
        header.ParseFromString(transaction.header)

        # The transaction signer is the player
        player = header.signer_pubkey

        try:
            # The payload is csv utf-8 encoded string
            name, action, space = transaction.payload.decode().split(",")
        except:
            raise InvalidTransaction("Invalid payload serialization")

        if name == "":
            raise InvalidTransaction("Name is required")

        if action == "":
            raise InvalidTransaction("Action is required")

        if action == 'CREATE':
            state_store[self._name] = {'State': 'NEW', 'Ships': self._ships}
        elif action == 'JOIN':
            game = state_store[self._name].copy()

            # If this is the first JOIN, set HashedBoard1 and Player1 in the
            # store.  if this is the second JOIN, set HashedBoard2 and
            # Player2 in the store.  Also, initialize TargetBoard1 and
            # TargetBoard2 as empty.
            if 'Player1' not in game:
                game['HashedBoard1'] = self._board
                size = len(self._board)
                game['TargetBoard1'] = [['?'] * size for _ in range(size)]
                game['Player1'] = self.OriginatorID
            else:
                game['HashedBoard2'] = self._board
                size = len(self._board)
                game['TargetBoard2'] = [['?'] * size for _ in range(size)]
                game['Player2'] = self.OriginatorID

                # Move to 'P1-NEXT' as both boards have been entered.
                game["State"] = 'P1-NEXT'

            state_store[self._name] = game
        elif self._action == 'FIRE':
            game = state_store[self._name].copy()

            # Reveal the previously targeted space
            if 'LastFireColumn' in game:
                if game['State'] == 'P1-NEXT':
                    target_board = 'TargetBoard2'
                else:
                    target_board = 'TargetBoard1'

                col = ord(game['LastFireColumn']) - ord('A')
                row = int(game['LastFireRow']) - 1
                if self._reveal_space != '-':
                    game[target_board][row][col] = 'H'
                else:
                    game[target_board][row][col] = 'M'

                # calculate number of hits for later determination
                # of win
                number_of_hits = sum(sum([1 if space == 'H' else 0
                                          for space in row])
                                     for row in game[target_board])
            else:
                number_of_hits = None

            # Update LastFireColumn and LastFireRow in the store so
            # they can be used in the next transaction.
            game['LastFireColumn'] = self._column
            game['LastFireRow'] = self._row

            # if the game has been won, change the State
            # to P1-WIN or P2-WIN as appropriate
            total_ship_spaces = sum([len(ship) for ship in game['Ships']])
            if number_of_hits is not None and \
                            total_ship_spaces == number_of_hits:
                if target_board == 'TargetBoard2':
                    game['State'] = 'P2-WIN'
                else:
                    game['State'] = 'P1-WIN'

            if game['State'] == 'P1-NEXT':
                game['State'] = 'P2-NEXT'
            elif game['State'] == 'P2-NEXT':
                game['State'] = 'P1-NEXT'

            state_store[self._name] = game
        else:
            raise InvalidTransaction(
                "invalid state: {}".format(state_store[self._name].copy))


def _is_win(board, letter):
    wins = ((1, 2, 3), (4, 5, 6), (7, 8, 9),
            (1, 4, 7), (2, 5, 8), (3, 6, 9),
            (1, 5, 9), (3, 5, 7))

    for win in wins:
        if (board[win[0] - 1] == letter
            and board[win[1] - 1] == letter
            and board[win[2] - 1] == letter):
            return True
    return False


def _game_data_to_str(board, state, player1, player2, name):
    board = list(board.replace("-", " "))
    out = ""
    out += "GAME: {}\n".format(name)
    out += "PLAYER 1: {}\n".format(player1[:6])
    out += "PLAYER 2: {}\n".format(player2[:6])
    out += "STATE: {}\n".format(state)
    out += "\n"
    out += "{} | {} | {}\n".format(board[0], board[1], board[2])
    out += "---|---|---\n"
    out += "{} | {} | {}\n".format(board[3], board[4], board[5])
    out += "---|---|---\n"
    out += "{} | {} | {}".format(board[6], board[7], board[8])
    return out


def _display(msg):
    n = msg.count("\n")

    if n > 0:
        msg = msg.split("\n")
        length = max(len(line) for line in msg)
    else:
        length = len(msg)
        msg = [msg]

    print("+" + (length + 2) * "-" + "+")
    for line in msg:
        print("+ " + line.center(length) + " +")
    print("+" + (length + 2) * "-" + "+")

File: families/battleship/src/handler.rs
// Copyright 2018 Bitwise IO, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use game;
use sawtooth_sdk::messages::processor::TpProcessRequest;
use sawtooth_sdk::processor::handler::{ApplyError, TransactionContext, TransactionHandler};
use serde_json;

pub struct BattleshipTransactionHandler {
    family_name: String,
    family_versions: Vec<String>,
    namespaces: Vec<String>,
}

impl Default for BattleshipTransactionHandler {
    fn default() -> Self {
        Self {
            family_name: "battleship".to_string(),
            family_versions: vec!["1.0".to_string()],
            namespaces: vec![game::get_battleship_prefix().to_string()],
        }
    }
}

impl BattleshipTransactionHandler {
    pub fn new() -> BattleshipTransactionHandler {
        BattleshipTransactionHandler {
            ..Default::default()
        }
    }

    /// Retrieves a single game by name
    fn get_game(
        context: &mut dyn TransactionContext,
        name: &str,
    ) -> Result<Option<game::Game>, ApplyError> {
        let address = game::get_battleship_address(name);
        let state = context.get_state_entry(&address)?;

        match state {
            Some(s) => serde_json::from_slice(s.as_slice()).map_err(|err| {
                ApplyError::InvalidTransaction(format!(
                    "Found existing state that couldn't be deserialized: {} (`{:?}`)",
                    err, s
                ))
            }),
            None => Ok(None),
        }
    }

    /// Stores a single game
    fn store_game(
        context: &mut dyn TransactionContext,
        name: &str,
        game: &game::Game,
    ) -> Result<(), ApplyError> {
        let address = game::get_battleship_address(name);

        let serialized = serde_json::to_string(game).map_err(|err| {
            ApplyError::InvalidTransaction(format!("Couldn't serialize game: {}", err))
        })?;

        context
            .set_state_entry(address, serialized.into())
            .map_err(|err| {
                ApplyError::InvalidTransaction(format!("Couldn't set new game state: {}", err))
            })
    }

    /// Handles CREATE action
    fn handle_create(
        context: &mut dyn TransactionContext,
        name: &str,
        ships: Vec<String>,
    ) -> Result<(), ApplyError> {
        match name.len() {
            1..=255 => {}
            len => Err(ApplyError::InvalidTransaction(format!(
                "Game name must be between 1 - 255 characters, {} is too many",
                len
            )))?,
        }

        if !name.chars().all(char::is_alphanumeric) {
            Err(ApplyError::InvalidTransaction(format!(
                "Game name must only contain characters A-Z, a-z or 0-9, found `{}`.",
                name
            )))?
        }

        if BattleshipTransactionHandler::get_game(context, name)?.is_some() {
            Err(ApplyError::InvalidTransaction(format!(
                "Game `{}` already exists.",
                name
            )))?
        }

        BattleshipTransactionHandler::store_game(
            context,
            name,
            &game::Game {
                ships,
                ..Default::default()
            },
        )
    }

    /// Handles JOIN action
    fn handle_join(
        context: &mut dyn TransactionContext,
        name: &str,
        player: String,
        board: Vec<Vec<String>>,
    ) -> Result<(), ApplyError> {
        let mut game = BattleshipTransactionHandler::get_game(context, name)?
            .ok_or_else(|| ApplyError::InvalidTransaction("Game not found.".into()))?;

        // Should be impossible given the checking we do down below, but here it is for completion
        if game.state != "NEW" {
            Err(ApplyError::InvalidTransaction(String::from("Game is full")))?
        }

        // Check that the board is a valid size
        if board.len() != 10 {
            Err(ApplyError::InvalidTransaction(String::from(
                "Invalid board size",
            )))?
        }

        // Check that each row is of length 10, and that each value in the row is a 128-byte long hash
        for row in &board {
            if row.len() != 10 {
                Err(ApplyError::InvalidTransaction(String::from(
                    "Invalid board size",
                )))?
            }

            for val in row.iter() {
                if val.len() != 128 {
                    Err(ApplyError::InvalidTransaction(String::from(
                        "Invalid board hash",
                    )))?
                }
            }
        }

        match (game.player_1.is_some(), game.player_2.is_some()) {
            // Nobody's joined
            (false, false) => {
                game.player_1 = Some(player);
                game.target_board_1 = board.iter().map(|_| vec!['?'; board.len()]).collect();
                game.hashed_board_1 = board;
            }
            // The first player has joined
            (true, false) => {
                if game.player_1 == Some(player.clone()) {
                    Err(ApplyError::InvalidTransaction(format!(
                        "Player {} is already in the game.",
                        player
                    )))?
                }
                game.player_2 = Some(player);
                game.target_board_2 = board.iter().map(|_| vec!['?'; board.len()]).collect();
                game.hashed_board_2 = board;
                game.state = "P1-NEXT".into();
            }
            // Something weird happened
            (false, true) => Err(ApplyError::InvalidTransaction(
                "Invalid game state, found P2 but no P1!".into(),
            ))?,
            // Go away, we're busy
            (true, true) => Err(ApplyError::InvalidTransaction(String::from("Game is full")))?,
        }

        BattleshipTransactionHandler::store_game(context, name, &game)
    }

    /// Handles FIRE action
    fn handle_fire(
        context: &mut dyn TransactionContext,
        name: &str,
        player: &str,
        column: String,
        row: String,
        reveal_space: Option<char>,
        reveal_nonce: &Option<String>,
    ) -> Result<(), ApplyError> {
        let mut game = BattleshipTransactionHandler::get_game(context, name)?
            .ok_or_else(|| ApplyError::InvalidTransaction("Game not found.".into()))?;

        let current_row = game::parse_row(&row)
            .ok_or_else(|| ApplyError::InvalidTransaction(format!("Invalid row: {}", row)))?;
        let current_col = game::parse_column(&column)
            .ok_or_else(|| ApplyError::InvalidTransaction(format!("Invalid column {}", column)))?;

        // Grab a mutable copy of the board that we'll update and reattach to `game`.
        let mut target_board = match game.state.as_ref() {
            // Everything checks out, return reference to target board
            "P1-NEXT" if game.player_1 == Some(player.to_string()) => game.target_board_2.clone(),
            "P2-NEXT" if game.player_2 == Some(player.to_string()) => game.target_board_1.clone(),

            // It's not their turn, reject the transaction
            "P1-NEXT" | "P2-NEXT" => Err(ApplyError::InvalidTransaction(format!(
                "It is not {}'s turn.",
                player
            )))?,

            // The game is either over or hasn't started, reject the transaction
            "P1-WIN" | "P2-WIN" => Err(ApplyError::InvalidTransaction(String::from(
                "Game complete!",
            )))?,
            "NEW" => Err(ApplyError::InvalidTransaction(String::from(
                "Game doesn't have enough players.",
            )))?,

            // Reject the impossible
            s => Err(ApplyError::InvalidTransaction(format!(
                "Invalid game state: {}",
                s
            )))?,
        };

        if target_board[current_row][current_col] != '?' {
            Err(ApplyError::InvalidTransaction(String::from(
                "Space already fired into.",
            )))?
        }

        // The current player should be revealing whether or not the other player's last
        // fire attempt was successful.
        match (
            &game.last_fire_column,
            &game.last_fire_row,
            reveal_space,
            reveal_nonce,
        ) {
            // Every turn after the first should hit this, unless the client tried giving
            // incomplete information.
            (Some(lfc), Some(lfr), Some(rs), Some(rn)) => {
                let last_row = game::parse_row(&lfr).ok_or_else(|| {
                    ApplyError::InvalidTransaction(format!("Invalid row {}", row))
                })?;

                let last_col = game::parse_column(&lfc).ok_or_else(|| {
                    ApplyError::InvalidTransaction(format!("Invalid column {}", column))
                })?;

                let space_hash = game::get_space_hash(rs, &rn);

                let hashed_board = match game.state.as_ref() {
                    "P1-NEXT" => &game.hashed_board_1,
                    "P2-NEXT" => &game.hashed_board_2,
                    s => Err(ApplyError::InvalidTransaction(format!(
                        "Invalid game state: {}",
                        s
                    )))?,
                };

                if hashed_board[last_row][last_col] != space_hash {
                    Err(ApplyError::InvalidTransaction(format!(
                        "Hash mismatch on reveal: {} != {}",
                        hashed_board[last_row][last_col], space_hash
                    )))?
                }

                match reveal_space {
                    Some('-') => target_board[last_row][last_col] = 'M',
                    Some(_) => target_board[last_row][last_col] = 'H',
                    None => {}
                }
            }

            // This should only happen on the first turn, since there's no previous fire info to send
            (None, None, None, None) => {}

            // All or nothing, reject any attempts to provide partial information
            _ => Err(ApplyError::InvalidTransaction(String::from(
                "Attempted to fire without revealing target",
            )))?,
        }

        let number_of_hits = target_board
            .iter()
            .map(|row| row.iter().filter(|item| **item == 'H').count())
            .sum::<usize>() as u32;

        game.last_fire_column = Some(column);
        game.last_fire_row = Some(row);

        let ships_length: u32 = game.ships.iter().map(|s| s.len() as u32).sum();

        match game.state.as_ref() {
            "P1-NEXT" => game.target_board_2 = target_board,
            "P2-NEXT" => game.target_board_1 = target_board,
            s => Err(ApplyError::InvalidTransaction(format!(
                "Invalid game state: {}",
                s
            )))?,
        }

        game.state = match (game.state.as_ref(), number_of_hits == ships_length) {
            ("P1-NEXT", true) => "P1-WIN".into(),
            ("P1-NEXT", false) => "P2-NEXT".into(),
            ("P2-NEXT", true) => "P2-WIN".into(),
            ("P2-NEXT", false) => "P1-NEXT".into(),
            (s, _) => Err(ApplyError::InvalidTransaction(format!(
                "Invalid state {}",
                s
            )))?,
        };

        BattleshipTransactionHandler::store_game(context, name, &game)
    }
}

impl TransactionHandler for BattleshipTransactionHandler {
    fn family_name(&self) -> String {
        self.family_name.clone()
    }

    fn family_versions(&self) -> Vec<String> {
        self.family_versions.clone()
    }

    fn namespaces(&self) -> Vec<String> {
        self.namespaces.clone()
    }

    fn apply(
        &self,
        request: &TpProcessRequest,
        context: &mut dyn TransactionContext,
    ) -> Result<(), ApplyError> {
        let payload = request.get_payload();
        let action: game::Action = serde_json::from_slice(payload).map_err(|err| {
            ApplyError::InvalidTransaction(format!("Error while parsing action: {}", err))
        })?;

        trace!("Parsed action: `{:?}`.", action);

        let player = request.get_header().get_signer_public_key().into();

        match action {
            game::Action::Create { name, ships } => {
                debug!("Creating game {} with {} ships.", name, ships.len());
                BattleshipTransactionHandler::handle_create(context, &name, ships)
            }
            game::Action::Join { name, board } => {
                debug!("Player {} is joining game {}.", player, name);
                BattleshipTransactionHandler::handle_join(context, &name, player, board)
            }
            game::Action::Fire {
                name,
                column,
                row,
                reveal_space,
                reveal_nonce,
            } => {
                debug!(
                    "Firing on {}{} in game {} for player {}.",
                    row, column, name, player
                );
                BattleshipTransactionHandler::handle_fire(
                    context,
                    &name,
                    &player,
                    column,
                    row,
                    reveal_space,
                    &reveal_nonce,
                )
            }
        }
    }
}
